dvc/stage.py

```diff
@@ -671,9 +669,18 @@ def dump(self):
         logger.debug(
             "Saving information to '{file}'.".format(file=relpath(fname))
         )
-        d = self.dumpd()
-        apply_diff(d, self._state)
-        dump_stage_file(fname, self._state)
+        state = self.dumpd()
+
+        # If we have stage file we apply diffs instead of overwriting it
+        # to preserve comments, string formatting and meta key.
+        if self._stage_text is not None:
+            saved_state = parse_stage_for_update(self._stage_text, fname)
```

Comment:
what is the difference between loaders? it looks like both of them return dicts?

Reply:
`ruamel.yaml` returns ordereddicts with comments stored in special attributes, it also uses str subclass to save string formatting, i.e. line breaks and indents manually added to command. 

So when you serialize that structure with `ruamel.yaml` into yaml string you get all those back.

Reply:
Thanks! But I still don't get it :) I can probably can take a deeper look into docs or something, but it's a bad sign and means that it is actually not easy to understand what was the motivation behind the decisions made. It means that someone else will try to remove and simplify this (I would) :). The question is - what is the difference between `parse_stage` and `parse_stage_for_update`. Why do we need to parse it second time? What's missing in the `parse_stage`? Comments, meta? Something else? Probably understand that can help me answer other question which I still don't have a clear answer for in my head. 

And please, let's think how can we rename, add tests, and/or add comments so that we better understand this.

Reply:
thanks for the update!